util/llbutil/authprovider/cloudstoredauthprovider/authprovider.go

```diff
@@ -0,0 +1,371 @@
+// Package cloudstoredauthprovider was forked from buildkit/session/auth/authprovider in order to allow using
+// registry credentials that are stored on a server rather than local filesystem.
+// This package is distributed under the original file's license, The Apache License, which is defined under
+// https://github.com/moby/buildkit/blob/7c3e9fdd48c867f48a07a80cde64cc2d578cb332/LICENSE
+
+package cloudstoredauthprovider
+
+import (
+	"context"
+	"crypto/ed25519"
+	"crypto/hmac"
+	"crypto/sha256"
+	"fmt"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	authutil "github.com/containerd/containerd/remotes/docker/auth"
+	remoteserrors "github.com/containerd/containerd/remotes/errors"
+	"github.com/docker/cli/cli/config"
+	"github.com/docker/cli/cli/config/configfile"
+	"github.com/docker/cli/cli/config/types"
+	"github.com/earthly/earthly/cloud"
+	"github.com/earthly/earthly/conslogging"
+	"github.com/earthly/earthly/util/llbutil/authprovider"
+	"github.com/moby/buildkit/session"
+	"github.com/moby/buildkit/session/auth"
+	"github.com/moby/buildkit/session/secrets"
+	"github.com/moby/buildkit/util/progress/progresswriter"
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/nacl/sign"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+const defaultExpiration = 60
+
+var ErrNoCredentialsFound = fmt.Errorf("no credentials found")
+
+type ProjectBasedAuthProvider interface {
+	AddProject(org, project string)
+}
+
+func NewCloudAuthProvider(cfg *configfile.ConfigFile, cloudClient cloud.Client, console conslogging.ConsoleLogger) session.Attachable {
```

Comment:
When calling this function, we'll be calling `cloudstoredauthprovider.NewCloudAuthProvider`, which stutters. I think this function can just be named `New`, or if we rename the package to `cloudauth`, maybe `NewProvider` (so it reads as `cloudauth.NewProvider`).